### DISCUSSION TITLE
Best practices for rate-limiting in durable functions?

### DISCUSSION THREAD

**Discussion by IanKemp** [Azure/azure-functions-durable-extension#1911] [Q&A]
I've been tasked with building an Azure durable function app that is aware of rate limits on HTTP endpoints. Essentially I'm going to have a timer trigger, that fires an orchestrator, that fires off an activity that gets a list of data. Using that list the orchestrator will then fire off a call to an activity for each item in that list; each activity will perform a single HTTP request.

The catch here is that the HTTP endpoints are rate-limited, but don't support status code 429 or the Retry-After header, so I need to be the one limiting my requests to them.

My current solution is for the orchestrator to batch the activities into sets of N items, where N is the maximum number of requests the endpoint allows, and run each batch as a fan-out. After a batch has completed, review the total time it took to complete said batch:
* If that duration is less than the rate limit duration, have the orchestrator sleep for the remainder of the rate limit duration
* Otherwise, run the next batch

The above seems like it should work, but it feels a little... *coarse*, and I'm worried there are pitfalls I'm not aware of. Am I overthinking this, or is there a simpler/more effective approach that I'm missing?

I've considered using Polly for this scenario, but I'm not certain if I even need it because I'm trying to control and limit the number of **durable functions activities** - which will indirectly accomplish rate-limiting the HTTP requests.

**Comment by ConnorMcMahon** [comment]
One option to consider is instead of using `CallActivityAsync()`, using `CallEntityAsync()`. Entities have similar flexibility to activity functions (i.e. they don't have to follow the orchestration rules). They allow you to persist state (such as a buffer of future HTTP Requests) and do things like schedule messages to wake up at a point in the future.

The psuedocode could look something like this.
```csharp
public RateLimiter : IRateLimiter
{
    public int MaxConcurrentRequests = 100;

    [JsonProperty("NextScheduledTime")]
    public DateTime? NextScheduleTime = {get; set;}

    [JsonProperty("BufferedRequests")]
    public List<HttpRequestData> BufferedRequests {get; set;}

    public Task ScheduleRequests(List<HttpRequestData> newRequests)
    {
        BufferedRequests.AddRange(newRequests)
         if (NextScheduledRequest == null) {
             await this.ProcessRequests();
         }
    }

    public Task ProcessRequests()
    {
         var batchToProcess = BufferedRequests.Take(MaxRequests);
        // async code to schedule http requests and measure time
        
         foreach (var response in httpResponses)
        {
             // Code to send notification back to orchestration that scheduled the request
            // that the request is completed and what the result was.
        }

         //Set to new state
        BufferedRequests = BufferedRequests.Skip(MaxRequests).ToList();
         if (rateLimitHit)
        {
             NextScheduleTime = nextTimeToProcessRequests;
        } else
        {
             NextScheduleTime = DateTime.UtcNow;
             this.ProcessRequests();
        }
    }
}

```

The orchestrator helper function would look something like this:

```csharp
public async Task<List<HttpResponseData> ScheduleHttpRequests(IDurableOrchestrationContext context,  List<HttpRequestData> requestsToSend)
{
     //NOTE: Since the rate limiter will handle that for me, can schedule all at once.
     await context.CallEntity(new EntityKey("RateLimiter", "api/path"), requestsToSend);

      // Code that waits for events from the entity that contain the responses. Can be one per http request or batched, depending on complexity of rate limiting code
}
```

Pros:

- Since the rate limiter entity exists across an entire taskhub, the rate limiting can happen _across orchestrations_ if you have other orchestrations relying on these APIs in the future
- All of your rate-limiting logic is in a single place, where your orchestration can just focus on business logic.

Cons:

- Entities are orchestrations under the covers, so it will only run on a single worker at a time, and each operation is processed sequentially.

**Comment by olitomlinson** [comment]
@IanKemp What granularity do the 3rd party endpoints rate limit at? And what is the max number of requests in that window?

i.e. Do they allow max 10 requests per second? 10 max request per minute? Per hour? Per rolling 24 hour period etc?

Do they allow concurrent requests, if so how many? or is it sequential-only? (you can't start the next request until the current in-flight request completes)

**Comment by sebastianburckhardt** [comment]
FYI see also the similar discussion here, which gives sample code for an entity acting as a semaphore: 
https://github.com/Azure/azure-functions-durable-extension/issues/2434

and we have sample code checked into the Netherite repo here, which was used for testing:
https://github.com/microsoft/durabletask-netherite/tree/main/test/PerformanceTests/Benchmarks/Semaphore
